lib/galaxy/jobs/runners/local.py

```diff
@@ -49,23 +49,53 @@ def __init__(self, app, nworkers):
 
         super().__init__(app, nworkers)
 
+    def get_local_gpu_count(self):
+        bash_command = "/bin/bash -c 'nvidia-smi --query-gpu=index --format=csv'"
+        sp = subprocess.Popen(bash_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = sp.communicate()
+        command_not_found = "command not found"
+        if command_not_found.encode() not in out and command_not_found.encode() not in err:
+            data = (out.decode('utf-8').splitlines())[1:] #eliminate first element, as it is the header
+            return len(data)
+        return 0
+
     def __command_line(self, job_wrapper):
         """ """
         command_line = job_wrapper.runner_command_line
 
         # slots would be cleaner name, but don't want deployers to see examples and think it
         # is going to work with other job runners.
         slots = job_wrapper.job_destination.params.get("local_slots") or os.environ.get("GALAXY_SLOTS")
+        cuda_visible_devices = job_wrapper.job_destination.params.get("cuda_visible_devices") or os.environ.get("CUDA_VISIBLE_DEVICES")
+        local_gpus = job_wrapper.job_destination.params.get("local_gpus") or os.environ.get("LOCAL_GPUS")
         if slots:
             slots_statement = f'GALAXY_SLOTS="{int(slots)}"; export GALAXY_SLOTS; GALAXY_SLOTS_CONFIGURED="1"; export GALAXY_SLOTS_CONFIGURED;'
         else:
             slots_statement = 'GALAXY_SLOTS="1"; export GALAXY_SLOTS;'
 
+        num_local_gpus = self.get_local_gpu_count()
+
+        if num_local_gpus > 0:
+            if cuda_visible_devices:
+                cvd_list = cuda_visible_devices.split(",")
+                if num_local_gpus < len(cvd_list):
+                    cvd_list = cvd_list[:num_local_gpus] #get the first num_local_gpus gpu indices if num_local_gpus is less than the length of visible devices specified
+                    cuda_visible_devices = ",".join(cvd_list)
+            elif local_gpus:
+                if num_local_gpus < int(local_gpus):
+                    local_gpus = num_local_gpus
+                cvd_list = range(local_gpus)
```

Comment:
My understanding was that the cuda visible device list needs to be a list of UUID strings, not just numbers: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-enumeration
Maybe using a technique like this to find the GPU with lowest utilization would do the trick? https://stackoverflow.com/a/69904259